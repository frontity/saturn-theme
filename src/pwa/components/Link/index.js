/* global window */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import NProgress from 'nprogress';
import { inject } from 'mobx-react';

class Link extends Component {
  static propTypes = {
    type: PropTypes.string.isRequired,
    id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
    page: PropTypes.number,
    context: PropTypes.shape({}),
    method: PropTypes.string,
    eventCategory: PropTypes.string,
    eventAction: PropTypes.string,
    children: PropTypes.node.isRequired,
    href: PropTypes.string.isRequired,
    routeChangeRequested: PropTypes.func.isRequired,
    sendEvent: PropTypes.func.isRequired,
  };

  static defaultProps = {
    page: null,
    method: 'push',
    context: null,
    eventCategory: null,
    eventAction: null,
  };

  constructor() {
    super();

    this.linkClicked = this.linkClicked.bind(this);
  }

  linkClicked(e) {
    // ignore click for new tab / new window behavior
    if (
      e.currentTarget.nodeName === 'A' &&
      (e.metaKey ||
        e.ctrlKey ||
        e.shiftKey ||
        (e.nativeEvent && e.nativeEvent.which === 2))
    )
      return;
    e.preventDefault();
    NProgress.configure({ showSpinner: false });
    NProgress.start();

    const {
      routeChangeRequested,
      sendEvent,
      type,
      id,
      page,
      context,
      method,
      eventCategory,
      eventAction,
    } = this.props;
    setTimeout(() => {
      sendEvent({ category: eventCategory, action: eventAction });

      routeChangeRequested({
        selectedItem: { type, id, page },
        context,
        method,
      });
    }, 100);
  }

  render() {
    const { children, href } = this.props;
    return React.cloneElement(children, { onClick: this.linkClicked, href });
  }
}

export default inject(
  ({ stores: { connection, analytics } }, { type, id, page }) => ({
    href: page
      ? connection.entity(type, id).pagedLink(page)
      : connection.entity(type, id).link,
    routeChangeRequested: connection.routeChangeRequested,
    sendEvent: analytics.sendEvent,
  }),
)(Link);
