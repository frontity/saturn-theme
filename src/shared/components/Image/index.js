/* eslint-disable jsx-a11y/anchor-is-valid, react/no-danger */
import React, { Component } from 'react';
import { bool, string } from 'prop-types';
import { inject } from 'mobx-react';
import styled from 'styled-components';
import IconImage from '../Icons/Image';

class Image extends Component {
  static propTypes = {
    isContent: bool,
    width: string, // CSS values
    height: string, // CSS values
    alt: string,
    src: string,
    srcSet: string,
    sizes: string,
    isAmp: bool.isRequired,
    hasPlaceholder: bool,
    objectFit: string,
  };

  static defaultProps = {
    isContent: false,
    width: 'auto',
    height: 'auto',
    alt: null,
    src: null,
    srcSet: null,
    sizes: null,
    hasPlaceholder: true,
    objectFit: 'cover',
  };

  componentDidMount() {
    window.document.lazyLoadInstance.update();
  }

  render() {
    const {
      alt,
      width,
      height,
      isContent,
      src,
      srcSet,
      sizes,
      isAmp,
      hasPlaceholder,
      objectFit,
    } = this.props;

    if (isAmp) {
      return (
        <Container isContent={isContent} styles={{ height, width }}>
          {src || srcSet ? (
            <amp-img
              alt={alt}
              src={src}
              srcSet={srcSet}
              sizes={sizes}
              layout="fill"
            />
          ) : null}
        </Container>
      );
    }

    return (
      <Container
        isContent={isContent}
        objectFit={objectFit}
        styles={{ height, width }}
      >
        {hasPlaceholder && (
          <Icon isContent={isContent} styles={{ height, width }}>
            <IconImage size={40} />
          </Icon>
        )}
        <span
          suppressHydrationWarning
          dangerouslySetInnerHTML={{
            __html:
              src || srcSet
                ? `<img
                    alt="alt"
                    class="lazy"
                    ${src && `data-src="${src}"`}
                    ${srcSet && `data-srcset="${srcSet}"`}
                    ${sizes && `data-sizes="${sizes}"`}
                  >`
                : null,
          }}
        />
      </Container>
    );
  }
}

export default inject(
  ({ stores: { connection, build } }, { id, isContent, width, height }) => {
    if (!id) return { isAmp: build.isAmp };

    const media = connection.entity('media', id);

    return {
      id,
      isAmp: build.isAmp,
      isContent: !!isContent,
      alt: media.alt,
      src: media.src,
      srcSet: media.srcSet,
      sizes: media.srcSet && `${parseInt(width, 10) || 100}vw`,
      width: width || '100vw',
      height:
        height ||
        `${100 * (media.original.height / media.original.width || 0.6)}vw`,
    };
  },
)(Image);

const Container = styled.span`
  display: ${({ isContent }) => (isContent ? 'block' : 'flex')};
  ${({ isContent }) => (isContent ? '' : 'align-items: stretch')};
  box-sizing: border-box;
  width: ${({ styles }) => styles.width};
  height: ${({ styles }) => styles.height};
  position: relative;
  margin: ${({ isContent }) => (isContent ? '15px 0' : '')};
  ${({ isContent }) => isContent && 'left: -15px'};

  span > img {
    ${({ isContent, styles }) =>
      isContent && styles.height === 'auto'
        ? 'position: static'
        : 'position: absolute'};
    width: 100%;
    height: 100%;
    object-fit: ${({ objectFit }) => objectFit};
    object-position: center;
    color: transparent;
    border: none;
    opacity: 0;
    transition: opacity ${({ theme }) => theme.transitionTime} ease-in;
  }

  amp-img > img {
    object-fit: cover;
  }

  img.loaded {
    opacity: 1;
  }
`;

const Icon = styled.span`
  position: absolute;
  top: 0;
  color: #bdbdbd;
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  ${({ isContent, styles }) =>
    isContent && styles.height === 'auto' ? 'z-index: -1' : ''};
`;
