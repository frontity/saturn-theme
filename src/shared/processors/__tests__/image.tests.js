/* eslint-disable react/prop-types */
import Image from '../../components/Image';
import Link from '../../../pwa/components/Link';
import processor from '../low/image';

const imageWithId = {
  type: 'element',
  component: 'img',
  props: {
    src: 'media/image.jpg',
    srcset: 'media/image.jpg',
    alt: 'image 1',
    'data-attachment-id': '23',
  },
};

const imageWithoutId = {
  type: 'element',
  component: 'img',
  props: {
    src: 'media/image.jpg',
    srcset: 'media/image.jpg',
    alt: 'image 1',
  },
};

const noImage = {
  type: 'element',
  component: 'div',
  props: { className: 'whatever' },
};

const media = {
  link: 'image.jpg',
  featured: {
    id: 23,
  },
  original: {
    width: 300,
    height: 250,
  },
};

const stores = {
  connection: {
    entity: () => media,
  },
};

const payload = {
  stores,
  item: {
    entity: { media },
  },
};

describe('H2R â€º Image processor', () => {
  test('does not pass test with invalid elements', () => {
    expect(processor.test(noImage)).toBeFalsy();
  });
  test('passes test with valid elements', () => {
    expect(processor.test(imageWithId)).toBeTruthy();
    expect(processor.test(imageWithoutId)).toBeTruthy();
  });
  test('process images with data-attachment-id', () => {
    const element = processor.process(imageWithId, payload);
    expect(element.component).toBe(Link);
    expect(element.children[0].children[0].component).toBe(Image);
    expect(element).toMatchSnapshot();
  });
  test('process images without data-attachment-id', () => {
    const element = processor.process(imageWithoutId, payload);
    expect(element.component).toBe(Image);
    expect(element).toMatchSnapshot();
  });
});
