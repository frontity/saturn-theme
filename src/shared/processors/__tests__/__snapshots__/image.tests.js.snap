// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`H2R › Image processor process images with data-attachment-id 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": null,
          "component": [Function],
          "props": Object {
            "id": 23,
            "isContent": true,
            "key": "23",
            "lazyloadContainerSelector": ".content",
          },
          "type": "element",
        },
      ],
      "component": "a",
      "props": Object {},
      "type": "element",
    },
  ],
  "component": [Function],
  "props": Object {
    "context": Object {
      "columns": Array [
        Array [
          Object {
            "id": 23,
            "type": "media",
          },
        ],
        Array [
          Object {
            "id": undefined,
            "type": "media",
          },
        ],
      ],
      "options": Object {
        "bar": "media",
      },
    },
    "eventAction": "open content media",
    "eventCategory": "Post",
    "id": 23,
    "type": "media",
  },
  "type": "element",
}
`;

exports[`H2R › Image processor process images without data-attachment-id 1`] = `
Object {
  "children": null,
  "component": [Function],
  "props": Object {
    "alt": "image 1",
    "height": "auto",
    "isContent": true,
    "key": "media/image.jpg",
    "lazyloadContainerSelector": ".content",
    "src": "media/image.jpg",
    "srcSet": null,
    "width": "100vw",
  },
  "type": "element",
}
`;
